metals/src/main/scala/scala/meta/internal/metals/WorkspaceLspService.scala

```diff
@@ -0,0 +1,1152 @@
+          .successful(Urls.openBrowser(githubNewIssueUrlCreator.buildUrl()))
+          .asJavaObject
+      case OpenBrowserCommand(url) =>
+        Future.successful(Urls.openBrowser(url)).asJavaObject
+      case ServerCommands.CascadeCompile() =>
+        foreachSeq(_.cascadeCompile(), ignoreValue = true)
+      case ServerCommands.CleanCompile() =>
+        foreachSeq(_.cleanCompile(), ignoreValue = true)
+      case ServerCommands.CancelCompile() =>
+        foreachSeq(_.cancelCompile(), ignoreValue = true)
+      case ServerCommands.PresentationCompilerRestart() =>
+        foreachSeq(_.restartCompiler(), ignoreValue = true)
+      case ServerCommands.GotoPosition(location) =>
+        Future {
+          languageClient.metalsExecuteClientCommand(
+            ClientCommands.GotoLocation.toExecuteCommandParams(
+              ClientCommands.WindowLocation(
+                location.getUri(),
+                location.getRange(),
+              )
+            )
+          )
+        }.asJavaObject
+
+      case ServerCommands.GotoSymbol(symbol) =>
+        Future {
+          for {
+            location <- folderServices.foldLeft(Option.empty[Location]) {
+              case (None, service) => service.getLocationForSymbol(symbol)
+              case (someLoc, _) => someLoc
+            }
+          } {
+            languageClient.metalsExecuteClientCommand(
+              ClientCommands.GotoLocation.toExecuteCommandParams(
+                ClientCommands.WindowLocation(
+                  location.getUri(),
+                  location.getRange(),
+                )
+              )
+            )
+          }
+        }.asJavaObject
+      case ServerCommands.GotoLog() =>
+        onCurrentFolder(
+          service =>
+            Future {
+              val log = service.folder.resolve(Directories.log)
+              val linesCount = log.readText.linesIterator.size
+              val pos = new lsp4j.Position(linesCount, 0)
+              val location = new Location(
+                log.toURI.toString(),
+                new lsp4j.Range(pos, pos),
+              )
+              languageClient.metalsExecuteClientCommand(
+                ClientCommands.GotoLocation.toExecuteCommandParams(
+                  ClientCommands.WindowLocation(
+                    location.getUri(),
+                    location.getRange(),
+                  )
+                )
+              )
+            },
+          "go to log",
+        ).asJavaObject
+
+      case ServerCommands.StartDebugAdapter(params) if params.getData != null =>
+        folderServices.find(s =>
+          params.getTargets().asScala.forall(s.supportsBuildTarget(_).isDefined)
+        ) match {
+          case Some(service) => service.startDebugProvider(params).asJavaObject
+          case None =>
+            // TODO:: show error message
+            Future.successful(()).asJavaObject
+        }
+      case ServerCommands.StartMainClass(params) if params.mainClass != null =>
```

Comment:
Looks like the four commands here have a very similar workflow. Could we maybe revert to previous behaviour of DebugProvider with a change that we would just run the first found main class? 

Or alternatively try to deduplicate the logic in these commands?